docs/services/overview-pages.js

```diff
@@ -51,39 +55,55 @@ function createPages(obj, slug) {
   });
 }
 
-const renderOverviewPages = (allPages, callback) => {
-  allPages.forEach(page => {
-    const { pages } = page;
-    if (!pages) return undefined;
-    callback(page);
-    return renderOverviewPages(pages, callback);
-  });
+const forEachNestedPage = (pages, callback) => {
+  pages
+    .filter(page => page.pages)
+    .forEach(page => {
+      callback({
+        ...page,
+        // keep pages nested only one level deep
+        pages: page.pages.map(p => _.omit(p, ["pages"])),
+      });
+      forEachNestedPage(page.pages, callback);
+    });
 };
 
-async function createOverviewPages(callback) {
+async function getOverviewPages() {
   const structure = {};
 
-  const files = await globby(path.join(__dirname, "../src/documentation/**/*.{yml,mdx}"));
+  // from these files we'll generate a list of pages containing nested pages in two stages
+  const files = await globby(path.join(DOCUMENTATION_PATH, "**/*.{yml,mdx}"));
+
+  // first stage
+  await Promise.all(
+    files.map(async file => {
+      const { dir, name } = path.parse(file);
+      const relativePath = omitNumbers(path.relative(DOCUMENTATION_PATH, path.join(dir, name)));
+
+      if (name === "meta") {
+        const { title, type, description } = yaml.load(await fsx.readFile(file, "utf-8"));
+        _.set(structure, relativePath.split(path.sep), { title, type, description });
+      } else {
+        const { data } = matter(await fsx.readFile(file, "utf-8"));
+        _.set(structure, relativePath.split(path.sep), data);
+      }
+    }),
+  );
 
-  files.forEach(file => {
-    const wholePath = omitNumbers(file)
-      .replace(/\.[^/.]+$/, "")
-      .split("/");
+  // second stage
+  const allPages = createPages(structure);
 
-    const ommitedPath = wholePath.slice(wholePath.indexOf("documentation") + 1, wholePath.length);
+  const overviewPages = [];
 
-    if (ommitedPath.includes("meta")) {
-      const { title, type, description } = yaml.load(fs.readFileSync(file));
-      _.set(structure, ommitedPath, { title, type, description });
-    } else {
-      const { data } = matter(fs.readFileSync(file, "utf-8"));
-      _.set(structure, ommitedPath, data);
-    }
+  // derive overview pages by recursively looping through pages that have nested pages
+  // this way we're essentially flattening the pages to be only one level deep
+  forEachNestedPage(allPages, page => {
+    overviewPages.push(page);
   });
```

Comment:
As above, I'm not sure this logic is necessary before the page is created, but if it's here, wouldn't it be a good use for reduce?

Reply:
I made this logic slightly more clear by inlining the recursive function at the place of invocation. I don't think this would be a good use case for `reduce` because `reduce` is best for deriving a single non-array value out of an array of elements. It's being overused in JavaScript because you can chain it to other array methods, but it often results in very unclear logic.